src/action.py

```diff
@@ -3,30 +3,69 @@
 import click
 import os
 from models.config import Config
+from collections import defaultdict
+from itertools import chain
 
 class ClickGlob:
-    def command(click_group, namespace):
-        def __glob_actions(namespace):
-            parts = ['/var/lib/adminware/tools',
-                     namespace, '*/config.yaml']
-            paths = glob.glob(os.path.join(*parts))
-            return list(map(lambda x: Action(x), paths))
+    def __glob_actions(namespace):
+        parts = ['/var/lib/adminware/tools',
+                 namespace, '*/config.yaml']
+        paths = glob.glob(os.path.join(*parts))
+        return list(map(lambda x: Action(x), paths))
 
+    def command(click_group, namespace):
         # command_func is either run_open or run_batch, what this is decorating
         def __command(command_func):
-            actions = __glob_actions(namespace)
+            actions = ClickGlob.__glob_actions(namespace)
             for action in actions:
                 action.create(click_group, command_func)
 
         return __command
 
+    def command_family(click_group, namespace):
+        def __command_family(command_func):
+            actions = ClickGlob.__glob_actions(namespace)
+            families = []
+            for action in actions:
+                if action.get_families(): families += [action.get_families()]
+            families_dict = __combine_dicts(families)
+            for key in families_dict.keys():
+                __create_option(click_group, command_func, key, families_dict[key])
+
+        # when want to combine the family dictionaries - while concatenating duplicate values
+        def __combine_dicts(dicts):
+            families_dict = defaultdict(list)
+            for single_dict in dicts:
+                for k, v in single_dict.items():
+                    families_dict[k].append(v)
+            # sort each command family
+            for key in families_dict:
+                families_dict[key].sort()
+            return families_dict
+
+        def __create_option(click_group, command_func, family, commands):
```

Comment:
The `__create_option` function should be define on the `Action`, because it creates the action. 
Also the next 8 lines of code duplicate:
https://github.com/alces-software/adminware/blob/042bfe7ccb5b017123f9405613b1c6732aad2450/src/action.py#L69-L73

Reply:
The issue with this, as I understand it, is that the `Action` class is inexorably tied to a single tool. It takes one `path` to its initializer and logically represents 1 command to be executed. The initial way I programmed this was to define a second class in the module, `ActionFamily`, which interfaced with `Action` however I removed this as i felt it hindered more than it helped comprehension - I can reinstate this if desired? Thinking about it there may have been better ways I could have implemented it, I'll draft up a version like this now & send it over to you for looking over

Why I didn't combine `create` and `create_option` into a single external helper method was the fact that `create` takes an argument of `arguments` but i suppose this could be worked around with a sizable amount of params? I'll put that in the updated version too

Reply:
Hmm it sounds like the existing functionality of `Action` is a special case of `ActionFamily` where there is only one action. Would there be a way to generate the `batch` and `open` commands using `ActionFamily`? That way we can trim down the code base